### PR TITLE
Workaround current Int Preview issue

### PR DESCRIPTION
**Pull request by dougbu**
- ignore `$vsInfo.installationVersion` unless it's valid

### CODE REVIEW

**Comment by mitchdenny** [comment]
I don't have enough context on what the behavior would be if we exclude the version parameter on the ```InitializeVisualStudioEnvironmentVariables``` function.

File: eng/common/tools.ps1 [hunk 1]
```diff
@@ -400,7 +400,9 @@ function InitializeVisualStudioMSBuild([bool]$install, [object]$vsRequirements =
   $vsInfo = LocateVisualStudio $vsRequirements
   if ($vsInfo -ne $null) {
     $vsInstallDir = $vsInfo.installationPath
-    $vsMajorVersion = $vsInfo.installationVersion.Split('.')[0]
+    if (Test-Path variable:vsInfo.installationVersion -and $vsInfo.installationVersion -match '^[\d.]+$') {
```
Comment (premun) [inline eng/common/tools.ps1:403]:
```suggestion
    // TODO: Remove this workaround (https://github.com/dotnet/arcade/issues/12726)
    if (Test-Path variable:vsInfo.installationVersion -and $vsInfo.installationVersion -match '^[\d.]+$') {
```

**Comment by nohwnd** [comment]
There is a problem upstream that causes this. This PR only solves the symptom. 

vswhere is unable to select the latest version reliably when you use -prerelease (at least not on my system). This is output of 10 tries:

PS S:\p\vstest2> 1..10 | % { S:\p\vstest2\.tools\vswhere\3.1.1\vswhere.exe -prerelease -latest -property installationPath } 
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise
C:\Program Files\Microsoft Visual Studio\2022\IntPreview
C:\Program Files\Microsoft Visual Studio\2022\IntPreview
C:\Program Files\Microsoft Visual Studio\2022\Preview
C:\Program Files\Microsoft Visual Studio\2022\IntPreview

This is because the installationVersion is gibbrish.

PS S:\p\vstest2> 1..10 | % { S:\p\vstest2\.tools\vswhere\3.1.1\vswhere.exe -prerelease -latest -property installationVersion } 
418Z7407
instanceId
instanceId
418Z7407
418Z7407
418Z7407
installDate

I am using version 3.1.1 of vswhere to make sure it was not fixed in the latest version, but the latest available without changing the dowload path is 2.7.1.

image


Maybe we can compare the versions ourselves, either from catalog.productSemanticVersion which seems to be returned correctly.

**Comment by nohwnd** [comment]
Seems to be the same for non-prerelease as well: 

1..10  | % { S:\p\vstest2\.tools\vswhere\2.2.7\vswhere.exe  -latest -property catalog.productSemanticVersion }
15.9.51+33027.88
17.4.3+33205.214
17.4.3+33205.214
17.4.3+33205.214
17.4.3+33205.214
17.4.3+33205.214
17.4.3+33205.214
17.4.3+33205.214
15.9.51+33027.88
15.9.51+33027.88

**Comment by premun** [comment]
@nohwnd I agree and would prefer to fix this at the root (in VS) than have this workaround. Just FYI, we're on a thread with VS Setup and they have been notified about this and will probably revert/fix the actual cause. More information [here](https://teams.microsoft.com/l/message/19:afba3d1545dd45d7b79f34c1821f6055@thread.skype/1677477650754?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=4d73664c-9f2f-450d-82a5-c2f02756606d&parentMessageId=1677477650754&teamName=.NET%20Core%20Eng%20Services%20Partners&channelName=First%20Responders&createdTime=1677477650754).

Maybe when they revert/fix this, we won't need to do anything here? Or do you propose this as a better solution anyway?

**Comment by nohwnd** [comment]
@premun we can quick fix this by ordering by catalog.productSemanticVersion it looks. Trying it now. Unfortunately there is no semanticVersion class in powershell 5 that is being used by the build engine. I can just split the version on - and + and compare that, which should be good enough, or at least deterministic.

**Comment by dougbu** [comment]
> vswhere is unable to select the latest version reliably when you use -prerelease (at least not on my system)

This is a side effect of the problem in the latest Int Preview version of VS.

> shouldn't we instead patch individual repos like this while that happens?

I created this PR in case the timeline was going to be longer. Closing because we've only got a day or so to wait.

**Pull request closed without merging**